File: 백준/Gold/2493. 탑/탑.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

class Data {
    int idx;
    int val;

    Data(int idx, int val) {
        this.idx = idx;
        this.val = val;
    }
}
public class Main {

    static int n;
    static int[] data, res;
    static Stack<Data> stack = new Stack<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine());
        data = new int[n];
        for (int i = 0; i < n; i++) {
            data[i] = Integer.parseInt(st.nextToken());
        }
        res = new int[n];

        for (int i = 0; i < n; i++) {
            if (stack.isEmpty()) {
                res[i] = 0;
                stack.push(new Data(i, data[i]));
            } else if (stack.peek().val > data[i]) {
                res[i] = stack.peek().idx+1;
                stack.push(new Data(i, data[i]));
            } else if (stack.peek().val < data[i]) {
                while (!stack.isEmpty()) {
                    if (stack.peek().val < data[i]) {
                        stack.pop();
                    } else {
                        res[i] = stack.peek().idx+1;
                        stack.push(new Data(i, data[i]));
                        break;
                    }
                }
                if (stack.isEmpty()) {
                    res[i] = 0;
                    stack.push(new Data(i, data[i]));
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(res[i]).append(" ");
        }
        System.out.println(sb);
    }
}

File: 백준/Gold/2493. 탑/탑.py
import sys
from collections import deque


n = int(input())
data = list(map(int, input().split()))
stack = []
res = []
for idx, val in enumerate(data):
    if len(stack) == 0: # 비어있으면
        stack.append((idx, val)) # 현재꺼 최신으로 저장
        res.append(0)
    elif stack[-1][1] > val: # 현재꺼보다 크면
        res.append(stack[-1][0]+1) # 마지막 인덱스 결과값에 저장
        stack.append((idx, val))
    elif stack[-1][1] < val: # 현재꺼보다 작으면
        while stack: # 스택이 존재할 때까지 반복
            if stack[-1][1] < val:
                stack.pop()
            else:
                res.append(stack[-1][0]+1)  # 해당 인덱스 저장하고
                stack.append((idx, val))
                break
        else:
            res.append(0)
            stack.append((idx,val))

for x in res:
    print(x, end = " ")